server/routes/createReaction.ts

```diff
@@ -1,32 +1,105 @@
+/* eslint-disable prefer-const */
+/* eslint-disable dot-notation */
+import { Response, NextFunction } from 'express';
 import lookupCommunityIsVisibleToUser from '../util/lookupCommunityIsVisibleToUser';
 import lookupAddressIsOwnedByUser from '../util/lookupAddressIsOwnedByUser';
-import { Response, NextFunction } from 'express';
 import { UserRequest } from '../types';
+import { NotificationCategories } from '../../shared/types';
+import { createCommonwealthUrl } from '../../shared/utils';
 
 const createReaction = async (models, req: UserRequest, res: Response, next: NextFunction) => {
   const [chain, community] = await lookupCommunityIsVisibleToUser(models, req.body, req.user, next);
   const author = await lookupAddressIsOwnedByUser(models, req, next);
+  const { reaction, comment_id, thread_id } = req.body;
 
-  if (!req.body.object_id) {
-    return next(new Error('Must provide object_id, address'));
+  if (!thread_id && !comment_id) {
+    return next(new Error('Must provide a comment or thread id'));
   }
-  if (!req.body.reaction) {
+  if (!reaction) {
     return next(new Error('Must provide text'));
   }
 
-  const reaction = await models.OffchainReaction.create(community ? {
-    community: community.id,
-    object_id: req.body.object_id,
-    reaction: req.body.reaction,
-    address_id: author.id,
-  } : {
-    chain: chain.id,
-    object_id: req.body.object_id,
-    reaction: req.body.reaction,
+  const options = {
+    reaction,
     address_id: author.id,
+  };
+
+  if (community) options['community'] = community.id;
+  else if (chain) options['chain'] = chain.id;
+  if (thread_id) options['thread_id'] = thread_id;
+  else if (comment_id) options['comment_id'] = comment_id;
+
+  let [ finalReaction, created ] = await models.OffchainReaction.findOrCreate({
+    where: options,
+    default: options,
+    include: [ models.Address]
+  });
+  if (created) finalReaction = await models.OffchainReaction.find({
+    where: options,
+    include: [ models.Address]
   });
 
-  return res.json({ status: 'Success', result: reaction.toJSON() });
+  let comment;
+  let cwUrl;
+  let root_type;
+  let proposal;
+  if (comment_id) {
+    comment = await models.OffchainComment.findByPk(Number(comment_id));
+    // Test on variety of comments to ensure root relation + type
+    const [prefix, id] = comment.root_id.split('_');
+    if (prefix === 'discussion') {
+      proposal = await models.OffchainThread.findOne({
+        where: { id }
+      });
+    } else if (prefix.includes('proposal') || prefix.includes('referendum')) {
```

Comment:
This seems like it could break easily to me, and I think we should just have the `else` condition search for Proposals.

Reply:
Will do. Thankfully, these workarounds should soon be a thing of the past!